From f13508fa702739aae1bfa85a7b25c1cacfedb2e9 Mon Sep 17 00:00:00 2001
From: llin4 <ling.lin@intel.com>
Date: Thu, 29 Mar 2012 14:48:34 +0800
Subject: dalvik: return positive hash code in Dalvik with Houdini

BZ: 29547

port from OTC https://otc-android.intel.com/gerrit/#/c/13089/

Make x86 version of Dalvik to always return positive hash code
when Houdini is enabled in the platform, to match
the ARM app's assumption on its native platform.
MiTalk fails without this fix.

Category: device-enablement
Domain: AOSP-Dalvik-Runtime
Origin: internal
Upstream-Candidate: no, needs rework

Change-Id: I22a218022ebba58e993d06a7f6a99c69b5d1c807
Orig-MCG-Change-Id: I417c39cd7a40871078f0597c3beffefcfec49189
Signed-off-by: llin4 <ling.lin@intel.com>
Signed-off-by: Qiming Shi <qiming.shi@intel.com>
Signed-off-by: Serguei Katkov <serguei.i.katkov@intel.com>
---
 Android.mk  |    9 ---------
 vm/Dvm.mk   |   10 ----------
 vm/Sync.cpp |   14 ++++----------
 3 files changed, 4 insertions(+), 29 deletions(-)

diff --git a/Android.mk b/Android.mk
index 4196c95..b2e2e15 100644
--- a/Android.mk
+++ b/Android.mk
@@ -23,15 +23,6 @@ ifeq ($(WITH_JIT),)
   endif
 endif
 
-WITH_POSITIVE_HASH := $(strip $(WITH_POSITIVE_HASH))
-ifeq ($(WITH_POSITIVE_HASH),)
-  ifeq ($(TARGET_ARCH),x86)
-    WITH_POSITIVE_HASH := true
-  else
-    WITH_POSITIVE_HASH := false
-  endif
-endif
-
 subdirs := $(addprefix $(LOCAL_PATH)/,$(addsuffix /Android.mk, \
 		libdex \
 		vm \
diff --git a/vm/Dvm.mk b/vm/Dvm.mk
index 09a1363..f4e1cb4 100644
--- a/vm/Dvm.mk
+++ b/vm/Dvm.mk
@@ -37,16 +37,6 @@ endif
 # Optional features.  These may impact the size or performance of the VM.
 #
 
-# Setting WITH_POSITIVE_HASH to true makes sure that hashCode()
-# always return positive values. Though this is not necessary
-# according to Java standard, many android apps make such
-# assumption because it's true on virtually every arm device.
-# On contrary, x86 devices may have negative hashCode.
-#
-ifeq ($(WITH_POSITIVE_HASH), true)
-  LOCAL_CFLAGS += -DWITH_POSITIVE_HASH
-endif
-
 # Houdini support
 LOCAL_CFLAGS += -DMTERP_NO_UNALIGN_64
 
diff --git a/vm/Sync.cpp b/vm/Sync.cpp
index dd110e0..2b4c025 100644
--- a/vm/Sync.cpp
+++ b/vm/Sync.cpp
@@ -1218,18 +1218,12 @@ void dvmThreadInterrupt(Thread* thread)
 #ifndef WITH_COPYING_GC
 u4 dvmIdentityHashCode(Object *obj)
 {
-#ifdef WITH_POSITIVE_HASH
     /*
-     * Clear the most significant bit of obj so that hashCode()
-     * always return positive values. Though this is not necessary
-     * according to Java standard, many android apps make such
-     * assumption because it's true on virtually every arm device.
-     * On contrary, x86 devices may have negative obj.
+     * The following assumes that objects are allocated at even boundaries, so
+     * the shift preserves uniqueness of hashCode() while guaranteeing a
+     * non-negative result (for the convenience of some applications,like MiTalk).
      */
-    return (u4)obj & 0x7fffffff;
-#else
-    return (u4)obj
-#endif
+    return (u4)(((u4)obj)>>1);
 }
 #else
 /*
-- 
1.7.4.1

